MapboxCoreNavigation/RouteController.swift

```diff
@@ -441,48 +502,109 @@ extension RouteController: CLLocationManagerDelegate {
             return
         }
+        
         let durationRemaining = routeProgress.durationRemaining
         
         isFindingFasterRoute = true
+        
+        print("[RouteController] Checking for faster/updated route...")
 
-        getDirections(from: location, along: routeProgress) { [weak self] (route, error) in
-            guard let strongSelf = self else {
-                return
-            }
+        getDirections(from: location, along: routeProgress) { [weak self] mostSimilarRoute, routes, error in
+            guard let self = self else { return }
             
             // Every request should reset the lastLocationDate, else we spam the server by calling this method every location update.
             // If the call fails, tough luck buddy! Then wait until the next interval before retrying
-            strongSelf.lastLocationDate = nil
+            self.lastLocationDate = nil
             
             // Also only do one 'findFasterRoute' call per time
-            strongSelf.isFindingFasterRoute = false
-
-            guard let route = route else {
+            self.isFindingFasterRoute = false
+            
+                
+            guard let route = mostSimilarRoute, let routes = routes else {
                 return
             }
-
-            guard let firstLeg = route.legs.first, let firstStep = firstLeg.steps.first else {
-                return
+            
+            self.applyNewRerouteIfNeeded(mostSimilarRoute: route, allRoutes: routes, currentUpcomingManeuver: currentUpcomingManeuver, durationRemaining: durationRemaining)
+        }
+    }
+    
+    func applyNewRerouteIfNeeded(mostSimilarRoute: Route, allRoutes: [Route], currentUpcomingManeuver: RouteStep, durationRemaining: TimeInterval) {
+        guard let firstLeg = mostSimilarRoute.legs.first, let firstStep = firstLeg.steps.first else {
+            return
+        }
+        
+        // Current and First step of old and new route should be significant enough of a maneuver before applying a faster route, so we don't apply the route just before a maneuver will occur
+        let significantFirstStep = firstStep.expectedTravelTime >= RouteControllerMediumAlertInterval && routeProgress.currentLegProgress.currentStepProgress.durationRemaining > RouteControllerMediumAlertInterval
+        
+        // Current maneuver should correspond to the next maneuver in the new route
+        let sameUpcomingManeuver = currentUpcomingManeuver == firstLeg.steps[1]
+        
+        // Check if the new route is faster by comparing the ETA to the current ETA. Should be 10% faster or more
+        let routeIsFaster = mostSimilarRoute.expectedTravelTime <= 0.9 * durationRemaining
+        
+        // Only check for alternatives if the user has plenty of time left on the route (10min+)
+        let userHasEnoughTimeOnRoute = self.routeProgress.durationRemaining > 600
```

Comment:
```suggestion
        let userHasEnoughTimeOnRoute = routeProgress.durationRemaining > 600
```

Reply:
This was already checked before, right?

Reply:
Theoretically this could change when a reroute call times out, but I'm fine with changing it 👍